pandas/tests/indexes/test_setops.py

```diff
@@ -95,3 +103,38 @@ def test_union_dtypes(left, right, expected):
     b = pd.Index([], dtype=right)
     result = (a | b).dtype
     assert result == expected
+
+
+def test_union_duplicate_index_subsets_of_each_other():
+    # GH: 31326
+    a = pd.Index([1, 2, 2, 3])
+    b = pd.Index([3, 3, 4])
+    expected = pd.Index([1, 2, 2, 3, 3, 4])
+
+    result = a.union(b)
+    tm.assert_index_equal(result, expected)
+    result = a.union(b, sort=False)
+    tm.assert_index_equal(result, expected)
+
+
+@pytest.mark.parametrize(
+    "func",
+    [
+        (Int64Index),
+        (Float64Index),
+        (DatetimeIndex),
+        (CategoricalIndex),
+        (TimedeltaIndex),
+    ],
+)
+def test_union_with_duplicate_index(func):
+    # GH: 36289
+    idx1 = func([1, 0, 0])
+    idx2 = func([0, 1])
+    expected = func([0, 0, 1])
```

Comment:
is there any need to mix/match types for idx1/idx2?  OK if that is out of scope

Reply:
Added one test for this case. Probably better to cover it. But is not really affected by the changes, because the result are bot indexes appended.